### PR TITLE
feat(network): create a bonded network

### PR DESCRIPTION
**Pull request by pdonias**
See vatesfr/xo-web#876

### CODE REVIEW

File: src/xapi-object-to-xo.js [hunk 1]
```diff
@@ -406,6 +406,7 @@ const TRANSFORMS = {
       type: 'PIF',
 
       attached: Boolean(obj.currently_attached),
+      bondMasterOf: obj.bond_master_of,
```
Comment (julien-f) [inline src/xapi-object-to-xo.js]: Bond objects are not available in the API, so it has very little value exposing this id.

Make it a boolean and rename it.

File: src/xapi/index.js [hunk 2]
```diff
@@ -1868,11 +1869,29 @@ export default class Xapi extends XapiBase {
...
 
+  async createBondedNetwork ({
+    name,
+    description,
+    pifIds,
+    mtu,
+    mac,
+    bond_mode
+  }) {
+    const network = await this.createNetwork({ name, description, mtu })
```
Comment (julien-f) [inline src/xapi/index.js]: Use object splat and spread to forward all unused params to `createNetwork()`.

File: src/xapi/index.js [hunk 3]
```diff
@@ -1868,11 +1869,29 @@ export default class Xapi extends XapiBase {
...
+    await this.call('Bond.create', network.$ref, map(pifIds, pifId => this.getObject(pifId).$ref), mac, bond_mode)
+
+    return network
+  }
+
   async deleteNetwork (networkId) {
     const network = this.getObject(networkId)
     await Promise.all(
-      mapToArray(network.$PIFs, (pif) => this.call('VLAN.destroy', pif.$VLAN_master_of.$ref))
+      mapToArray(network.$PIFs, pif => pif.$VLAN_master_of && this.call('VLAN.destroy', pif.$VLAN_master_of.$ref))
```
Comment (julien-f) [inline src/xapi/index.js]: Use an intermediary variable for `$VLAN_master_of`.

File: src/xapi/index.js [hunk 4]
```diff
@@ -1868,11 +1869,29 @@ export default class Xapi extends XapiBase {
...
     )
+    const bond = !isEmpty(network.$PIFs) && network.$PIFs[0].bond_master_of[0]
```
Comment (julien-f) [inline src/xapi/index.js]: Intermediary variable for pifs.

File: src/xapi/index.js [hunk 5]
```diff
@@ -1868,11 +1869,29 @@ export default class Xapi extends XapiBase {
...
+    if (bond) {
+      await this.call('Bond.destroy', bond)
```
Comment (julien-f) [inline src/xapi/index.js]: You should probably use `map()` here.

File: src/xapi/index.js [hunk 6]
```diff
@@ -1868,11 +1869,29 @@ export default class Xapi extends XapiBase {
...
+  async createBondedNetwork ({
+    name,
+    description,
+    pifIds,
+    mtu,
+    mac,
+    bond_mode
+  }) {
+    const network = await this.createNetwork({ name, description, mtu })
+    await this.call('Bond.create', network.$ref, map(pifIds, pifId => this.getObject(pifId).$ref), mac, bond_mode)
```
Comment (julien-f) [inline src/xapi/index.js]: We are creating a bond for a single host and XAPI should create all the other bonds automatically.

Add a comment explaining that and that we are not sure about it.

File: src/xapi/index.js (hunk 3 above)
Comment (julien-f) [inline src/xapi/index.js]: I fear that we may try to delete the same VLAN multiple times.

Comment (julien-f) [inline src/xapi/index.js]: Better use this simplification rule: `obj.$prop.$ref` →`obj.prop`.

File: src/xapi/index.js (hunk 4 above)
Comment (julien-f) [inline src/xapi/index.js]: Do we need to also iterate over all PIFs?

File: src/api/network.js [hunk 7]
```diff
@@ -61,7 +54,7 @@ createBonded.params = {
     }
   },
   mtu: { type: ['integer', 'string'], optional: true },
-  bond_mode: { type: 'string' }
+  bondMode: { type: 'string', pattern: new RegExp(`^(${getBondModes().join('|')})$`) }
```
Comment (julien-f) [inline src/api/network.js]: Maybe add a comment saying there are no enums in schema-inspector.

File: src/xapi/index.js [hunk 8]
```diff
@@ -1870,28 +1871,33 @@ export default class Xapi extends XapiBase {
...
+    await this.call('Bond.create', network.$ref, map(pifIds, pifId => this.getObject(pifId).$ref), mac, bondMode)
 
     return network
   }
 
   async deleteNetwork (networkId) {
     const network = this.getObject(networkId)
+    const pifs = network.$PIFs
+
+    const vlans = uniq(mapToArray(pifs, pif => pif.VLAN_master_of))
```
Comment (julien-f) [inline src/xapi/index.js]: What does it do when there `VLAN_master_of` is *empty*.

File: src/xapi/index.js (hunk 6 above)
Comment (slindaqiu) [inline src/xapi/index.js]: Where is the definition of 'Bond.create'?

Comment (olivierlambert) [inline src/xapi/index.js]: @slindaqiu http://xapi-project.github.io/xen-api/classes/bond.html

**Pull request merged**